File: src/main/java/de/tectoast/emolga/utils/ConfigManager.java
package de.tectoast.emolga.utils;

import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.dexquiz.DexQuizTip;
import de.tectoast.emolga.utils.config.ConfigSource;

import java.util.function.Function;

public enum ConfigManager {
    DEXQUIZ(s -> {
        if (s.equals("totalbudget")) return Command.DEXQUIZ_BUDGET;
        return DexQuizTip.valueOf(s).getDefaultPrice();
    });

    private final Function<String, Object> defaultValue;
    private final ConfigSource configSource;

    ConfigManager(Function<String, Object> defaultValue) {
        this(defaultValue, ConfigSource.JSON);
    }

    ConfigManager(Function<String, Object> defaultValue, ConfigSource configSource) {
        this.defaultValue = defaultValue;
        this.configSource = configSource;
    }

    public Object getValue(long gid, String key) {
        return configSource.get(gid, name().toLowerCase(), key, () -> defaultValue.apply(key));
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/ConfigManager.kt
package de.tectoast.emolga.utils

import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.dexquiz.DexQuizTip
import de.tectoast.emolga.utils.config.ConfigSource
import java.util.*
import java.util.function.Function

enum class ConfigManager(private val defaultValue: Function<String, Any>) {
    DEXQUIZ(Function<String, Any> { s: String ->
        if (s == "totalbudget") return@Function Command.DEXQUIZ_BUDGET
        DexQuizTip.valueOf(s).defaultPrice
    });

    private val configSource: ConfigSource = ConfigSource.JSON

    fun getValue(gid: Long, key: String): Any {
        return configSource[gid, name.lowercase(Locale.getDefault()), key, { defaultValue.apply(key) }]
    }
}